internal/helm/helm.go

```diff
@@ -0,0 +1,168 @@
+	// ChartName:   "/path/to/stable/etcd-operator",			Local directory
+	// ChartName:   "/path/to/stable/etcd-operator.tar.gz",			Local archive
+	// ChartName:   "http://helm.whatever.com/repo/etcd-operator.tar.gz",	Remote archive
+)
+
+type ChartParameters struct {
+	models.AppRef                       // Application: name & namespace
+	Cluster       *kubernetes.Cluster   // Cluster to talk to.
+	Chart         string                // Chart to use for deployment
+	ImageURL      string                // Application Image
+	Username      string                // User causing the (re)deployment
+	Instances     int32                 // Number Of Desired Replicas
+	StageID       string                // Stage ID for ImageURL
+	Owner         metav1.OwnerReference // App CRD Owner Information
+	Environment   models.EnvVariableMap // App Environment
+	Services      []string              // Bound Services (list of names)
+	Routes        []string              // Desired application routes
+	Start         *int64                // Nano-epoch of deployment. Optional. Used to force a restart, even when nothing else has changed.
+}
+
+func Remove(cluster *kubernetes.Cluster, logger logr.Logger, app models.AppRef) error {
+	client, err := getHelmClient(cluster, logger, app.Namespace)
+	if err != nil {
+		return err
+	}
+
+	return client.UninstallReleaseByName(names.ReleaseName(app.Name))
+}
+
+func Deploy(logger logr.Logger, parameters ChartParameters) error {
+
+	// YAML string - TODO ? Use unstructured as intermediary to
+	// marshal yaml from ? Instead of direct generation of a
+	// string ?
+
+	serviceNames := `[]`
+	if len(parameters.Services) > 0 {
+		serviceNames = fmt.Sprintf(`["%s"]`, strings.Join(parameters.Services, `","`))
+	}
+
+	environment := `[]`
+	if len(parameters.Environment) > 0 {
+		// TODO: Simplify the chain of conversions. Single `AsYAML` ?
+		environment = fmt.Sprintf(`[ %s ]`, strings.Join(parameters.Environment.List().Assignments(),
+			","))
+	}
+
+	routesYaml := "~"
+	if len(parameters.Routes) > 0 {
+		rs := []string{}
+		for _, desired := range parameters.Routes {
+			r := routes.FromString(desired)
+			rs = append(rs, fmt.Sprintf(`{"id":"%s","domain":"%s","path":"%s"}`,
+				strings.ReplaceAll(r.String(), "/", "."),
+				r.Domain, r.Path))
+		}
+		routesYaml = fmt.Sprintf(`[%s]`, strings.Join(rs, `,`))
+	}
+
+	ingress := "~"
+	name := viper.GetString("ingress-class-name")
+	if name != "" {
+		ingress = name
+	}
+
+	start := ""
+	if parameters.Start != nil {
+		start = fmt.Sprintf(`start: "%d"`, *parameters.Start)
+	}
+
+	yamlParameters := fmt.Sprintf(`
+epinio:
+  appName: "%[10]s"
+  appUID: "%[2]s"
+  env: %[7]s
+  imageURL: "%[4]s"
+  ingress: %[11]s
+  replicaCount: %[1]d
+  routes: %[8]s
+  services: %[6]s
+  stageID: "%[3]s"
+  tlsIssuer: "%[12]s"
+  username: "%[5]s"
+  %[9]s
+`, parameters.Instances,
+		parameters.Owner.UID,
```

Comment:
Given [we now remove with helm](https://github.com/epinio/epinio/pull/1225/files#diff-44a311773090a2abd70d10d9f026c5d98e1a8f88f37381e50f8b4a30fe784854R240), I wonder if we still need to set owner references on the various resources. After all, if someone deletes the AppCRD without calling `helm uninstall` on the app release, the resulting state will be invalid from the helm's perspective. We can clean this up in another PR.

Reply:
True that. Would definitely simplify the chart. There was already an issue with ownership, the certs for the ingress' is now owned by the app, not the ingress as before.